File: client/src/utils/crc16.js
const CRC16 = {
  calculate(data) {
    const bytes = typeof data === 'string' ? new TextEncoder().encode(data) : data;
    let crc = 0xffff;
    
    for (let i = 0; i < bytes.length; i++) {
      crc ^= bytes[i] << 8;
      for (let j = 0; j < 8; j++) {
        if (crc & 0x8000) {
          crc = (crc << 1) ^ 0x1021;
        } else {
          crc <<= 1;
        }
        crc &= 0xffff;
      }
    }
    
    return crc;
  },

  verify(data, expectedCrc) {
    const calculated = this.calculate(data);
    return calculated === expectedCrc;
  }
};

export default CRC16;

File: server/src/main/java/com/minecraftmanager/websocket/util/CRC16.java
package com.minecraftmanager.websocket.util;

public class CRC16 {
    private static final int POLYNOMIAL = 0x1021;
    private static final int PRESET = 0xFFFF;

    public static short calculate(byte[] data) {
        int crc = PRESET;
        
        for (byte b : data) {
            crc ^= (b & 0xFF) << 8;
            for (int i = 0; i < 8; i++) {
                if ((crc & 0x8000) != 0) {
                    crc = (crc << 1) ^ POLYNOMIAL;
                } else {
                    crc <<= 1;
                }
                crc &= 0xFFFF;
            }
        }
        
        return (short) crc;
    }

    public static boolean verify(byte[] data, short expectedCrc) {
        short calculated = calculate(data);
        return calculated == expectedCrc;
    }
}